tools/permission_prompt_test.py

```diff
@@ -5,220 +5,189 @@
     me, se = pty.openpty()
     mi, si = pty.openpty()
     fdmap = {mo: 'stdout', me: 'stderr', mi: 'stdin'}
 
+    timeout_exact = time.time() + timeout
     p = subprocess.Popen(
         cmd, bufsize=1, stdin=si, stdout=so, stderr=se, close_fds=True)
     os.write(mi, bytes_input)
 
-    timeout = .04  # seconds
+    select_timeout = .04  #seconds
     res = {'stdout': b'', 'stderr': b''}
     while True:
-        ready, _, _ = select.select([mo, me], [], [], timeout)
+        ready, _, _ = select.select([mo, me], [], [], select_timeout)
         if ready:
             for fd in ready:
                 data = os.read(fd, 512)
                 if not data:
                     break
                 res[fdmap[fd]] += data
-        elif p.poll() is not None:  # select timed-out
+        elif p.poll() is not None or time.time(
+        ) > timeout_exact:  # select timed-out
             break  # p exited
     for fd in [si, so, se, mi, mo, me]:
         os.close(fd)  # can't do it sooner: it leads to errno.EIO error
     p.wait()
     return p.returncode, res['stdout'], res['stderr']
 
 
+def wrap_test(test_name, test_method, *argv):
+    sys.stdout.write(test_name + " ... ")
+    try:
+        test_method(*argv)
+        print green_ok()
+    except AssertionError:
+        print red_failed()
+        raise
+
+
 class Prompt(object):
-    def __init__(self, deno_exe):
+    def __init__(self, deno_exe, test_types):
         self.deno_exe = deno_exe
+        self.test_types = test_types
 
-    def run(self,
-            arg,
-            bytes_input,
-            allow_read=False,
-            allow_write=False,
-            allow_net=False,
-            allow_env=False,
-            allow_run=False,
-            no_prompt=False):
+    def run(self, arg, bytes_input, args=None):
         "Returns (return_code, stdout, stderr)."
-        cmd = [self.deno_exe, PERMISSIONS_PROMPT_TEST_TS, arg]
-        if allow_read:
-            cmd.append("--allow-read")
-        if allow_write:
-            cmd.append("--allow-write")
-        if allow_net:
-            cmd.append("--allow-net")
-        if allow_env:
-            cmd.append("--allow-env")
-        if allow_run:
-            cmd.append("--allow-run")
-        if no_prompt:
-            cmd.append("--no-prompt")
+        if args is None:
+            args = []
+        cmd = [self.deno_exe, PERMISSIONS_PROMPT_TEST_TS, arg] + args
         return tty_capture(cmd, bytes_input)
 
     def warm_up(self):
         # ignore the ts compiling message
-        self.run('needsWrite', b'', allow_write=True)
-
-    def test_read_yes(self):
-        code, stdout, stderr = self.run('needsRead', b'y\n')
+        self.run('needsWrite', b'', ["--allow-write"])
+
+    def test(self):
+        for test_type in self.test_types:
+            test_name_base = "test_" + test_type
+            wrap_test(test_name_base + "_allow_flag", self.test_allow_flag,
```

Comment:
I think this would be clearer as a second for loop.

Reply:
It might look cleaner, but I think from a debugging perspective it's not worth removing that much explicitness from a test. This would likely only save a few lines. I will add some comments to `wrap_test`.

Reply:
How so? if you debug then you can access the variable name.